sdk/internal/perf/perf.go

```diff
@@ -0,0 +1,270 @@
+	fmt.Println("Running `Setup`")
+
+	ctx, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.Setup(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// runTest takes care of the semantics of running a single iteration. It returns the number of times the test ran as an int, the exact number
+// of seconds the test ran as a float64, and any errors.
+func runTest(p PerfTest, c chan runResult) {
+	defer wg.Done()
+	fmt.Println("Beginning `Run`")
+	ctx, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	// If we are using the test proxy need to set up the in-memory recording.
+	if TestProxy != "" {
+		// First request goes through in Live mode
+		setRecordingMode("live")
+		err := p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+
+		// 2nd request goes through in Record mode
+		setRecordingMode("record")
+		start(p.GetMetadata(), nil)
+		err = p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+		stop(p.GetMetadata(), nil)
+
+		// All ensuing requests go through in Playback mode
+		setRecordingMode("playback")
+		start(p.GetMetadata(), nil)
+	}
+
+	start := time.Now()
+	count := 0
+	for time.Since(start).Seconds() < float64(Duration) {
+		err := p.Run(ctx)
+		if err != nil {
+			c <- runResult{count: 0, timeInSeconds: 0.0, err: err}
+		}
+		count += 1
+	}
+
+	elapsed := time.Since(start).Seconds()
+
+	// Stop the proxy now
+	stop(p.GetMetadata(), nil)
+	setRecordingMode("live")
+	c <- runResult{count: count, timeInSeconds: elapsed, err: nil}
+}
+
+// runTearDown takes care of the semantics for tearing down a single iteration of a performance test.
+func runTearDown(p PerfTest) error {
+	fmt.Println("Running `Teardown`")
+
+	_, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.TearDown(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func runGlobalTearDown(p PerfTest) error {
+	fmt.Println("Running `GlobalTeardown`")
+	_, cancel := getLimitedContext(time.Duration(TimeoutSeconds) * time.Second)
+	defer cancel()
+
+	err := p.GlobalTearDown(context.Background())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type runResult struct {
+	count         int
+	timeInSeconds float64
+	err           error
+}
+
+func runPerfTest(p PerfTest) error {
+	err := runGlobalSetup(p)
+	if err != nil {
+		return err
+	}
+
+	totalCount := 0
+	totalTime := 0.0
+
+	runSetup(p)
+
+	var channels []chan runResult
+
+	for idx := 0; idx < Parallel; idx++ {
+		wg.Add(1)
+		c := make(chan runResult, 1) // Create a buffered channel
+		channels = append(channels, c)
+		go runTest(p, c)
+	}
+
+	wg.Wait()
+
+	// Get the results from the channels
+	for idx, channel := range channels {
+		result := <-channel
+		if err != nil {
+			return err
+		}
+		printIteration(idx, result.timeInSeconds, result.count)
```

Comment:
Other frameworks only print the total across all threads, not results per thread.

Reply:
Removed